File: solutions/2026-04-05/cpp/2026-04-05-robot-return-to-origin.cpp
// Problem: Robot Return to Origin
// Link: https://leetcode.com/problems/robot-return-to-origin/
// Approach: We can simulate the robot's movement by keeping track of its current x and y coordinates.
// We initialize both x and y to 0. Then, we iterate through the given moves string.
// For each move:
// - If the move is 'U', we increment y.
// - If the move is 'D', we decrement y.
// - If the move is 'R', we increment x.
// - If the move is 'L', we decrement x.
// After processing all moves, we check if both x and y are back to 0.
// Time complexity: O(n), where n is the length of the moves string, because we iterate through the string once.
// Space complexity: O(1), because we only use a few variables to store the coordinates, regardless of the input size.

#include <string>
#include <iostream>

class Solution {
public:
    bool judgeCircle(std::string moves) {
        // Initialize the robot's position to the origin (0, 0)
        int x = 0;
        int y = 0;

        // Iterate through each move in the input string
        for (char move : moves) {
            // Update the coordinates based on the move
            if (move == 'U') {
                y++; // Move up
            } else if (move == 'D') {
                y--; // Move down
            } else if (move == 'R') {
                x++; // Move right
            } else if (move == 'L') {
                x--; // Move left
            }
        }

        // Check if the robot has returned to the origin (0, 0)
        return x == 0 && y == 0;
    }
};

/*
// Example Usage:
int main() {
    Solution sol;

    std::string moves1 = "UD";
    std::cout << "Moves: " << moves1 << ", Returns to origin: " << (sol.judgeCircle(moves1) ? "true" : "false") << std::endl; // Expected: true

    std::string moves2 = "LL";
    std::cout << "Moves: " << moves2 << ", Returns to origin: " << (sol.judgeCircle(moves2) ? "true" : "false") << std::endl; // Expected: false

    std::string moves3 = "RRDD";
    std::cout << "Moves: " << moves3 << ", Returns to origin: " << (sol.judgeCircle(moves3) ? "true" : "false") << std::endl; // Expected: false

    std::string moves4 = "LDRRLRUULR";
    std::cout << "Moves: " << moves4 << ", Returns to origin: " << (sol.judgeCircle(moves4) ? "true" : "false") << std::endl; // Expected: false

    return 0;
}
*/

File: solutions/2026-04-05/python/2026-04-05-robot-return-to-origin.py
```python
# Problem: Robot Return to Origin
# Link: https://leetcode.com/problems/robot-return-to-origin/
#
# Approach:
# We can simulate the robot's movement by keeping track of its current x and y coordinates.
# Initialize x and y to 0.
# Iterate through the `moves` string.
# For each move:
#   - If the move is 'U', increment y.
#   - If the move is 'D', decrement y.
#   - If the move is 'R', increment x.
#   - If the move is 'L', decrement x.
# After processing all moves, check if both x and y are 0.
# If both are 0, the robot returned to the origin, so return True. Otherwise, return False.
#
# Alternative Approach (more efficient for this specific problem):
# Instead of simulating step-by-step, we can count the occurrences of each move.
# The robot returns to the origin if and only if the number of 'U' moves equals the number of 'D' moves,
# AND the number of 'L' moves equals the number of 'R' moves.
# This is because 'U' and 'D' cancel each other out vertically, and 'L' and 'R' cancel each other out horizontally.
#
# Time Complexity: O(N), where N is the length of the `moves` string.
# We iterate through the string once to count the moves.
# Space Complexity: O(1), as we only use a few variables to store counts, regardless of input size.

class Solution:
    def judgeCircle(self, moves: str) -> bool:
        # Initialize counts for each direction.
        # We can use variables or a dictionary/Counter.
        # Using variables is slightly more efficient for a fixed set of characters.
        up_count = 0
        down_count = 0
        left_count = 0
        right_count = 0

        # Iterate through each move in the input string.
        for move in moves:
            if move == 'U':
                up_count += 1
            elif move == 'D':
                down_count += 1
            elif move == 'L':
                left_count += 1
            elif move == 'R':
                right_count += 1

        # Check if the net movement in the vertical direction is zero.
        # This means the number of 'U' moves must equal the number of 'D' moves.
        vertical_return = (up_count == down_count)

        # Check if the net movement in the horizontal direction is zero.
        # This means the number of 'L' moves must equal the number of 'R' moves.
        horizontal_return = (left_count == right_count)

        # The robot returns to the origin if and only if both vertical and horizontal
        # movements result in a return to the origin.
        return vertical_return and horizontal_return

# Example Usage:
# solver = Solution()
# print(solver.judgeCircle("UD"))  # Output: True
# print(solver.judgeCircle("LL"))  # Output: False
# print(solver.judgeCircle("RRDD")) # Output: False
# print(solver.judgeCircle("LDRRLRUULR")) # Output: False
```